File: devops-scm-provider/devops-scm-provider-git/devops-scm-provider-gitee/src/main/java/com/tencent/devops/scm/provider/git/gitee/GiteeRefService.java
package com.tencent.devops.scm.provider.git.gitee;

import com.tencent.devops.scm.api.RefService;
import com.tencent.devops.scm.api.pojo.BranchListOptions;
import com.tencent.devops.scm.api.pojo.Change;
import com.tencent.devops.scm.api.pojo.Commit;
import com.tencent.devops.scm.api.pojo.CommitListOptions;
import com.tencent.devops.scm.api.pojo.ListOptions;
import com.tencent.devops.scm.api.pojo.Reference;
import com.tencent.devops.scm.api.pojo.ReferenceInput;
import com.tencent.devops.scm.api.pojo.TagListOptions;
import com.tencent.devops.scm.api.pojo.repository.ScmProviderRepository;
import com.tencent.devops.scm.api.pojo.repository.git.GitScmProviderRepository;
import com.tencent.devops.scm.provider.git.gitee.auth.GiteeTokenAuthProviderAdapter;
import com.tencent.devops.scm.sdk.common.enums.SortOrder;
import com.tencent.devops.scm.sdk.gitee.GiteeApi;
import com.tencent.devops.scm.sdk.gitee.GiteeApiFactory;
import com.tencent.devops.scm.sdk.gitee.GiteeBranchesApi;
import com.tencent.devops.scm.sdk.gitee.GiteeTagApi;
import com.tencent.devops.scm.sdk.gitee.enums.GiteeBranchOrderBy;
import com.tencent.devops.scm.sdk.gitee.pojo.GiteeBranch;
import com.tencent.devops.scm.sdk.gitee.pojo.GiteeCommitCompare;
import com.tencent.devops.scm.sdk.gitee.pojo.GiteeCommitDetail;
import com.tencent.devops.scm.sdk.gitee.pojo.GiteeTag;
import com.tencent.devops.scm.sdk.gitee.pojo.GiteeTagDetail;
import java.util.List;
import java.util.stream.Collectors;

public class GiteeRefService implements RefService {

    private final GiteeApiFactory apiFactory;

    public GiteeRefService(GiteeApiFactory apiFactory) {
        this.apiFactory = apiFactory;
    }

    @Override
    public void createBranch(ScmProviderRepository repository, ReferenceInput input) {

    }

    @Override
    public Reference findBranch(ScmProviderRepository repository, String name) {
        GitScmProviderRepository repo = (GitScmProviderRepository) repository;
        GiteeApi giteeApi = apiFactory.fromAuthProvider(GiteeTokenAuthProviderAdapter.get(repo.getAuth()));
        // 构建请求接口类
        GiteeBranchesApi branchesApi = giteeApi.getBranchesApi();
        GiteeBranch branch = branchesApi.getBranch(repo.getProjectIdOrPath(), name);
        return GiteeObjectConverter.convertBranches(branch);
    }

    @Override
    public List<Reference> listBranches(ScmProviderRepository repository, BranchListOptions opts) {
        GitScmProviderRepository repo = (GitScmProviderRepository) repository;
        GiteeApi giteeApi = apiFactory.fromAuthProvider(GiteeTokenAuthProviderAdapter.get(repo.getAuth()));
        // 构建请求接口类
        GiteeBranchesApi branchesApi = giteeApi.getBranchesApi();
        List<GiteeBranch> branches = branchesApi.getBranches(repo.getProjectIdOrPath());
        // 结果转化
        return branches.stream().map(GiteeObjectConverter::convertBranches).collect(Collectors.toList());
    }

    @Override
    public void createTag(ScmProviderRepository repository, ReferenceInput input) {

    }

    @Override
    public Reference findTag(ScmProviderRepository repository, String name) {
        GitScmProviderRepository repo = (GitScmProviderRepository) repository;
        GiteeApi giteeApi = apiFactory.fromAuthProvider(GiteeTokenAuthProviderAdapter.get(repo.getAuth()));
        // 构建请求接口类
        GiteeTagApi tagApi = giteeApi.getTagApi();
        GiteeTagDetail tag = tagApi.getTags(repo.getProjectIdOrPath(), name);
        return GiteeObjectConverter.convertTag(tag);
    }

    @Override
    public List<Reference> listTags(ScmProviderRepository repository, TagListOptions opts) {
        GitScmProviderRepository repo = (GitScmProviderRepository) repository;
        GiteeApi giteeApi = apiFactory.fromAuthProvider(GiteeTokenAuthProviderAdapter.get(repo.getAuth()));
        // 构建请求接口类
        GiteeTagApi tagApi = giteeApi.getTagApi();
        List<GiteeTag> tags = tagApi.getTags(
                repo.getProjectIdOrPath(),
                opts.getPage(),
                opts.getPageSize(),
                GiteeBranchOrderBy.valueOf(opts.getOrderBy()),
                SortOrder.valueOf(opts.getSort())
        );
        return tags.stream().map(GiteeObjectConverter::convertTag).collect(Collectors.toList());
    }

    @Override
    public Commit findCommit(ScmProviderRepository repository, String ref) {
        GitScmProviderRepository repo = (GitScmProviderRepository) repository;
        GiteeApi giteeApi = apiFactory.fromAuthProvider(GiteeTokenAuthProviderAdapter.get(repo.getAuth()));
        GiteeCommitDetail commit = giteeApi.getCommitApi().getCommit(repo.getProjectIdOrPath(), ref);
        return GiteeObjectConverter.convertCommit(commit);
    }

    @Override
    public List<Commit> listCommits(ScmProviderRepository repository, CommitListOptions opts) {
        return null;
    }

    @Override
    public List<Change> listChanges(ScmProviderRepository repository, String ref, ListOptions opts) {
        GitScmProviderRepository repo = (GitScmProviderRepository) repository;
        GiteeApi giteeApi = apiFactory.fromAuthProvider(GiteeTokenAuthProviderAdapter.get(repo.getAuth()));
        GiteeCommitDetail commit = giteeApi.getCommitApi().getCommit(
                repo.getProjectIdOrPath(),
                ref
        );
        return GiteeObjectConverter.convertChange(commit);
    }

    @Override
    public List<Change> compareChanges(
            ScmProviderRepository repository,
            String source,
            String target,
            ListOptions opts
    ) {
        GitScmProviderRepository repo = (GitScmProviderRepository) repository;
        GiteeApi giteeApi = apiFactory.fromAuthProvider(GiteeTokenAuthProviderAdapter.get(repo.getAuth()));
        GiteeCommitCompare commitCompare = giteeApi.getFileApi().commitCompare(
                repo.getProjectIdOrPath(),
                source,
                target,
                null
        );
        return GiteeObjectConverter.convertCompare(commitCompare);
    }
}

File: devops-scm-provider/devops-scm-provider-git/devops-scm-provider-gitee/src/main/kotlin/com/tencent/devops/scm/provider/git/gitee/GiteeRefService.kt
package com.tencent.devops.scm.provider.git.gitee

import com.tencent.devops.scm.api.RefService
import com.tencent.devops.scm.api.pojo.BranchListOptions
import com.tencent.devops.scm.api.pojo.Change
import com.tencent.devops.scm.api.pojo.Commit
import com.tencent.devops.scm.api.pojo.CommitListOptions
import com.tencent.devops.scm.api.pojo.ListOptions
import com.tencent.devops.scm.api.pojo.Reference
import com.tencent.devops.scm.api.pojo.ReferenceInput
import com.tencent.devops.scm.api.pojo.TagListOptions
import com.tencent.devops.scm.api.pojo.repository.ScmProviderRepository
import com.tencent.devops.scm.api.pojo.repository.git.GitScmProviderRepository
import com.tencent.devops.scm.sdk.common.enums.SortOrder
import com.tencent.devops.scm.sdk.gitee.GiteeApiFactory
import com.tencent.devops.scm.sdk.gitee.enums.GiteeBranchOrderBy

class GiteeRefService(private val apiFactory: GiteeApiFactory) : RefService {

    override fun createBranch(repository: ScmProviderRepository, input: ReferenceInput) {
        // 暂未实现
    }

    override fun findBranch(repository: ScmProviderRepository, name: String): Reference {
        return GiteeApiTemplate.execute(repository, apiFactory) { repo, giteeApi ->
            val branchesApi = giteeApi.getBranchesApi()
            val branch = branchesApi.getBranch(repo.projectIdOrPath, name)
            GiteeObjectConverter.convertBranches(branch)
        }
    }

    override fun listBranches(
        repository: ScmProviderRepository,
        opts: BranchListOptions
    ): List<Reference> {
        return GiteeApiTemplate.execute(repository, apiFactory) { repo, giteeApi ->
            val branchesApi = giteeApi.getBranchesApi()
            val branches = branchesApi.getBranches(repo.projectIdOrPath)
            branches.map { GiteeObjectConverter.convertBranches(it) }
        }
    }

    override fun createTag(repository: ScmProviderRepository, input: ReferenceInput) {
        // 暂未实现
    }

    override fun findTag(repository: ScmProviderRepository, name: String): Reference {
        return GiteeApiTemplate.execute(repository, apiFactory) { repo, giteeApi ->
            val tagApi = giteeApi.getTagApi()
            val tag = tagApi.getTags(repo.projectIdOrPath, name)
            GiteeObjectConverter.convertTag(tag)
        }
    }

    override fun listTags(
        repository: ScmProviderRepository,
        opts: TagListOptions
    ): List<Reference> {
        return GiteeApiTemplate.execute(repository, apiFactory) { repo, giteeApi ->
            val tagApi = giteeApi.getTagApi()
            val tags = tagApi.getTags(
                repo.projectIdOrPath,
                opts.page,
                opts.pageSize,
                opts.orderBy?.let { GiteeBranchOrderBy.valueOf(it) },
                opts.sort?.let { SortOrder.valueOf(it) }
            )
            tags.map { GiteeObjectConverter.convertTag(it) }
        }
    }

    override fun findCommit(repository: ScmProviderRepository, ref: String): Commit {
        return GiteeApiTemplate.execute(repository, apiFactory) { repo, giteeApi ->
            val commit = giteeApi.getCommitApi().getCommit(repo.projectIdOrPath, ref)
            GiteeObjectConverter.convertCommit(commit)
        }
    }

    override fun listCommits(
        repository: ScmProviderRepository,
        opts: CommitListOptions
    ): List<Commit> {
        // 暂未实现
        return emptyList()
    }

    override fun listChanges(
        repository: ScmProviderRepository,
        ref: String,
        opts: ListOptions
    ): List<Change> {
        return GiteeApiTemplate.execute(repository, apiFactory) { repo, giteeApi ->
            val commit = giteeApi.getCommitApi().getCommit(repo.projectIdOrPath, ref)
            GiteeObjectConverter.convertChange(commit)
        }
    }

    override fun compareChanges(
        repository: ScmProviderRepository,
        source: String,
        target: String,
        opts: ListOptions
    ): List<Change> {
        return GiteeApiTemplate.execute(repository, apiFactory) { repo, giteeApi ->
            val commitCompare = giteeApi.getFileApi().commitCompare(
                repo.projectIdOrPath,
                source,
                target,
                null
            )
            GiteeObjectConverter.convertCompare(commitCompare)
        }
    }
}
